### ISSUE TITLE
wish: Publish to Debian ?

### ISSUE THREAD

**Issue by rzr** [markdownlint/markdownlint#310]
Maybe I should report to debian tracker,
but did anyone try to package this tool ?

Ping me here or at 
https://mastodon.social/@rzr/103809689316855902

**Comment by jaymzh** [comment]
It is not built as a debian package. We would welcome someone doing that!

**Comment by brlin-tw** [comment]
Note that there's a snap package available: [Install mdl for Linux using the Snap Store | Snapcraft](https://snapcraft.io/mdl)

**Comment by nbehrnd** [comment]
The suggest to offer a .deb about mdl has been filed as suggestion to Debian,
[ticket 1025320](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1025320).

**Comment by rzr** [comment]
Thanks may I suggest to try to join existing team:

https://wiki.debian.org/Teams/Ruby

**Comment by nbehrnd** [comment]
@rzr Thank you for the provision of an entry link.   At present, I understood
the instructions on https://wiki.debian.org/Teams/Ruby/Packaging/gem2deb are
suitable because there wasn't a `mdl.deb` (yet).

Running Debian 12/bookworm (branch testing), early steps appeared promising to
the untrained eye until the guide proceeds with a `gbp import-dsc` (about centre
of the web page) and reaches a road block (log attached below).  Apparently, it
is not something an `apt-get install gdb` is able to resolve *per sé* -- do you
have an idea if `gbp` belongs to an other package `apt` can extend/install?

[2022-12-05_mdl2deb.log](https://github.com/markdownlint/markdownlint/files/10157903/2022-12-05_mdl2deb.log)

**Comment by nbehrnd** [comment]
@mivok If obstacles preparing a suitable .deb for Debian are resolved, I would
like to submit this packaged version to Debian's repositories.  An intermediate
step in preparation is accessing the repository on https://rubygems.org/gems/mdl
you maintain.  In this perspective, I would like both to inform you about this
intent as well as to ask you *in advance ahead* doing so for permission.

**Comment by jaymzh** [comment]
@nbehrnd - `gbp` vs `gdb` ? But anyway the package is `git-buildpackage`:

```shell
$ apt-file search bin/gbp
gbutils: /usr/bin/gbplot                  
git-buildpackage: /usr/bin/gbp
lxqt-build-tools: /usr/share/lxqt-build-tools/bin/gbp-buildpackage
lxqt-build-tools: /usr/share/lxqt-build-tools/bin/gbp-import-orig
```

**Comment by mivok** [comment]
Hi, @nbehrnd if you mean read only access, no permission is required, although I appreciate the heads up. If you mean read write access to publish new versions of the gem, I'm curious why this is needed, but I'll defer to other maintainers as I'm no longer the primary maintainer of markdownlint.

**Comment by nbehrnd** [comment]
@jaymzh It indeed was a misspell on my side (`gdb` is a debugger some use
to improve Fortran programs).  With your hint (I had to install the package),
the generation of a .deb can be continued.

**Comment by nbehrnd** [comment]
@mivok `gem2deb` accesses https://rubygems.org/gems/mdl to collect data needed
to construct a .deb package.  This read-only access of this repository is to be
complemented by additional manual input and some corrections in the template
provides.  Since it would be my first .deb to deposit there, I have to walk
through their their detailed documentation; e.g., the adjustment of parameters
for dependencies and their pre-dependencies could be an issue here.  This might
take a little time to digest.

**Comment by nbehrnd** [comment]
@rzr The .deb below is a preliminary one (i.e., does not fulfill all of Debian's
quality standards) and probably worked (Xubuntu 22.04.1 LTS) only because
markdownlint's dependencies already were met.  However, running a

```shell
sudo dpgk -i ruby-mdl_0.12.0-1_all.deb
```

eventually yielded a functional `mdl` to the shell (in my case, bash) previously
absent.

[ruby-mdl_0.12.0-1_all.deb.zip](https://github.com/markdownlint/markdownlint/files/10167183/ruby-mdl_0.12.0-1_all.deb.zip)
[2022-12-31_Debian_build.zip](https://github.com/markdownlint/markdownlint/files/10338076/2022-12-31_Debian_build.zip)

**Comment by nbehrnd** [comment]
@psyomn, @jaymzh @mivok (retrieved from https://rubygems.org/gems/mdl)
@jaymzh, @NaomiReeves, @bwann, @DavidAnson  (retrieved from `MAINTAINERS.md`)

Mark Harrison is listed as one of the past maintainers of markdownlint.  On the other hand, the script-based preparation of a .deb package of `markdownlint` yields two files `control` and `copyright` (drafts attached below) extracting some of the information compiled, including his email address as `upstream-contact`.  As I understand, this however should be someone *deeply* familiar with the program .and. willing to reply to bug reports about `markdownlint` turned-in via Debian's `reportbug`; i.e. preferentially one of the current maintainers.  Who of you would like to step-in, what contact address should I add instead of Mark's?

[control.txt](https://github.com/markdownlint/markdownlint/files/10222473/control.txt)
[copyright.txt](https://github.com/markdownlint/markdownlint/files/10222474/copyright.txt)

**Comment by jaymzh** [comment]
You can use mine. Phil at ipom dot comOn Dec 13, 2022 2:02 PM, nbehrnd ***@***.***> wrote:
@psyomn, @jaymzh @mivok (retrieved from https://rubygems.org/gems/mdl)
@jaymzh, @NaomiReeves, @bwann, @DavidAnson  (retrieved from MAINTAINERS.md)
Mark Harrison is listed as one of the past maintainers of markdownlint.  On the other hand, the script-based preparation of a .deb package of markdownlint yields two files control and copyright (drafts attached below) extracting some of the information compiled, including his email address as upstream-contact.  As I understand, this however should be someone deeply familiar with the program .and. willing to reply to bug reports about markdownlint turned-in via Debian's reportbug; i.e. preferentially one of the current maintainers.  Who of you would like to step-in, what contact address should I add instead of Mark's?
control.txt
copyright.txt

**Comment by rzr** [comment]
@nbehrnd may I suggest to aslo commit your debian files in a branch , there i can double check if this can be built from sources

**Comment by nbehrnd** [comment]
@rzr Thank you for offering assistance here.  For one, I created a new branch [here](https://github.com/nbehrnd/markdownlint/tree/mdl2debian_nbehrnd) which includes the current stage of building in a .zip archive stamped by today, [2022-12-15].  This branch not intended to push for a merge here to markdownlint equally contains the two helper scripts I wrote to ensure all contributors of markdownlint (up and including version 0.12.0) are mentioned in the `copyright` file generated the preparation with `gem2deb` generates.

But I'm stuck.  After filing the ITP recommended ([entry](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1026114)), as understood so far, I should follow now [this other guide](https://wiki.debian.org/Teams/Ruby/Packaging/ITP). So far, the 

> Once you added your package as a git repo to * [gitorious debian-diaspora](https://gitorious.org/debian-diaspora) you should add the info about the git-repository to the bug.

appears to me like something is missing (which could be on my side).  Do you have experience submitting a (Ruby) application to Debian .and. can you point to a more recent guide (the current version of https://wiki.debian.org/Teams/Ruby/Packaging/ITP was revised by 2020-11-11, maybe interfaces changed since them [which is my speculation])?

**Comment by rzr** [comment]
next step is request for sponsorship but before  please commit debian/* files somewhere  and finally try to upload to https://mentors.debian.net/

**Comment by nbehrnd** [comment]
Sorry, I misunderstood you, the previous .zip archive including the .deb build wasn't the format anticipated by you.

For the moment, what `gem2deb` yielded in `~/Build/ruby-packaging/ruby-mdl/debian/` is stashed [here](https://github.com/nbehrnd/markdownlint/tree/mdl2debian_02/debian).  I'm now reading the documentation https://mentors.debian.net/intro-maintainers/ provides.

**Comment by rzr** [comment]
ok it;s a good start see some of my comments at:

https://github.com/nbehrnd/markdownlint/commit/5871a36c928b0db2d12925be0fc75d8a0ed45b2b

please revise your history only have debian/* changes in commits history, you can rebase on latest release tag and keep the minimal changes needed and meanwhile try to reach debian ruby team , you can mention me if it helps

**Comment by DavidAnson** [comment]
@jaymzh You should probably remove my name from the maintainers list at some point. I am subscribed to this repo and pay attention to what's going on, but I am not in a position to make fixes or publish releases.

**Comment by jaymzh** [comment]
@DavidAnson - I think that was added as a stop-gap to not have only me when I took over. I'll remove you now.

**Comment by psyomn** [comment]
Same here, please remove me from owners -- thank you!

**Comment by jaymzh** [comment]
@psyomn you're not listed as a current maintainer, only a past maintainer

**Comment by psyomn** [comment]
https://rubygems.org/gems/mdl <- I am listed as an owner here.  If you can, please rescind my access.

**Comment by jaymzh** [comment]
Done!

**Comment by nbehrnd** [comment]
@rzr With `ruby-mdl` as package name, `markdownlint` currently seeks review and sponsorship by an already approved Debian maintainer ([public entry](https://mentors.debian.net/package/ruby-mdl/)).

**Comment by rzr** [comment]
ok good next step please try to fix all mistakes reported at:

https://mentors.debian.net/package/ruby-mdl/

ie: write manpage etc

**Comment by nbehrnd** [comment]
@rzr Early I recognized `gem2deb` does not prepare a man page, so I wrote one based on `markdownlint`'s help, too. Puzzling to me is this observation: if it is absent, `lintian` (rightfully) points this out as missing.  From [section 5.15](https://www.debian.org/doc/manuals/maint-guide/dother.en.html#manpage) `/debian/mdl.1` appeared as correct storage for the submission  yesterday.  However, now `lintian` once issues a warning (section `ruby-mdl`) *as absent* with a yellow flag, and simultaneously issues a green flag (understood as *good to advance*) *as present* in `ruby-mdl source`.  Do you recommend to deposit the identical same file twice?  Or, are two different man pages necessary?

Reading the definition of the next warning (`maintainer-also-in-uploaders`) [here](https://lintian.debian.org/tags/maintainer-also-in-uploaders) appears to me as a remain of a previously useful/necessary criterion to check, and today *this is not as severe as is an absent man page*.

**Comment by nbehrnd** [comment]
By [2023-01-05 Thu 05:36 UTC] and thanks to sponsor [kilobyte](https://qa.debian.org/developer.php?login=kilobyte), `ruby-mdl` has been added to https://ftp-master.debian.org/new.html.  To quote Linux Debian's [documentation](https://wiki.debian.org/NewQueue)

> "When a new package is uploaded to Debian for the first time, it enters the ["NEW queue"](http://ftp-master.debian.org/new.html), and awaits approval from an [FTP Team member](https://wiki.debian.org/Teams/FTPMaster)."

Because I do not know when this next advance takes place (and an entry on https://tracker.debian.org eventually can be be accessed), my comment by [2022-12-06 Tue] in this thread still includes an unofficial .deb built by [2022-12-31 Sat].

For a visual overview about the 50-percentile, 90-percentile, and 99-percentile, see [Processing Times for the NEW Queue](https://people.debian.org/~roehling/new_queue/).

**Comment by rzr** [comment]
good job

https://ftp-master.debian.org/new/ruby-mdl_0.12.0-1.html


now wait for ftp masters and track 

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1026114

**Comment by nbehrnd** [comment]
Independent to the initial bug report (to Debian) I filed which informs me, I'm going to equally visit «the other» pages, too. Meanwhile, sponsoring mentor Adam Borowski suggested how to improve the current packaging further (no impact on eventually running the application, nor change of the source code of `markdownlint`; yet additional improvement to «debianize» the files assisting Debian).

**Comment by nbehrnd** [comment]
@jaymzh On [rubygems](https://rubygems.org/gems/mdl), runtime dependencies list e.g., 

```
mixlib-cli ~> 2.1, >= 2.1.1
```

So far, I assumed a release stamped as version `2.1.1` relies on prior publication of version `2.1`.  In this perspective, wouldn't either 

```
mixlib-cli ~> 2.1
```

or

```
mixlib-cli ~>= 2.1.1
```

*alone* suffice?  (As by today [2023-01-06 Fri], [mixlib-cli](https://rubygems.org/gems/mixlib-cli) is distributed as version `2.1.8`.)

**Comment by jaymzh** [comment]
`~> 2.1` would allow 2.1.0 and I think we needed at least 2.1.1

I don't know what `~>=` is

**Comment by nbehrnd** [comment]
@jaymzh According to rubygems's documentation, section [Declaring dependencies](https://guides.rubygems.org/patterns/#declaring-dependencies), this notation is introduced as *pessimistic version constraint* (see a sub section of this title, about half through the page).  Later, the documentation equally refers to this as [twiddle-wakka](http://robots.thoughtbot.com/post/2508037841/twiddle-wakka).  To quote from the second reference

> Do you know Ruby’s pessimistic operator? It looks like this:
> 
>  ```
>  ~>
>  ```
>
> You’ve seen it in some Gemfiles. Here’s an example
>
>  ```
> gem 'rails', '~> 3.0.3'
> gem 'thin',  '~> 1.1'
> ```
> 
> `~> 3.0.3` means that when you `bundle install`, you’ll get the highest-released gem version of `rails` between the range `>= 3.0.3` and `< 3.1`.
> 
> `~> 1.1` means that when you `bundle install`, you’ll get the highest-released gem version of `thin` between the range `>= 1.1` and `< 2.0`.

Because rubygems current provision of `mixlib-cli` is about version `2.1.8`, I think the dependency definition for `mdl` were fine with

```
mixlib-cli ~> 2.1
```

alone.

**Comment by jaymzh** [comment]
Er, I know what `~>` does. You used a `~>=` operator, which afaik, does not exist.

And no, `mixlib-cli ~> 2.1` won't work. As I said, that'd allow 2.1.0 and I we needed at least 2.1.1.

**Comment by jaymzh** [comment]
BTW there's even this specific example in the docs you reference:

```
If you want to allow use of newer backwards-compatible versions but need a specific bug fix you can use a compound requirement:

    # gemspec
    spec.add_runtime_dependency 'library', '~> 2.2', '>= 2.2.1'

    # bundler
    gem 'library', '~> 2.2', '>= 2.2.1'
```

which is exactly what we are doing.

**Comment by nbehrnd** [comment]
To write 

```
mixlib-cli ~>= 2.1.1
```

was an error on my side, the tilde was a sign too many.  Neither `pessimistic version constraint`, nor `twiddle-wakka` were previously known to me.  It is by Ruby I encounter the underlying concept, and name.  They extend definitions already familiar like `mixlib-config >= 2.2.1, < 4` to stay within a range of versions released.

In this perspective

```
mixlib-cli ~> 2.1, >= 2.1.1
```

refers to any version of `mixlib-cli 2.1.X` as long as it is either `mixlib-cli 2.1.1`, or higher.

**Comment by jaymzh** [comment]
Yes, exactly. :)

**Comment by nbehrnd** [comment]
By now ([2023-01-08 Sun 19:00] (CET)), there is an entry for `ruby-mdl` on Debian: https://tracker.debian.org/pkg/ruby-mdl for the branch of `unstable`.  By scale of Debian's quality control, however, there still is work to be done (paraphrasing my mentor «A `release 0.12.0-1` only prepends `release 0.12.0-2`.»)

**Comment by nbehrnd** [comment]
As an intermediate update I would like to let you know `markdownlint` as package [`ruby-mdl`](https://tracker.debian.org/pkg/ruby-mdl) entered Linux Debian 12/bookworm, branch `testing`.  Aside that this attracted some Debian relatives (so far, [repology](https://repology.org/project/mdl-markdownlint/packages) identified Devuan, Raspbian, Ubuntu 23.04), this transition introduces the application into the next Debian release anticipated for Q1/2023 (time table [freeze policy](https://release.debian.org/testing/freeze_policy.html)).

Because the character of `markdownlint` is more the one of an application, eventually, I would like to rename the package *on Debian's repositories* as `markdownlint` rather than `ruby-mdl`.  I would appreciate if one of you is aware of a documentation/a tuto to complement my question to the Debian mentors (https://lists.debian.org/debian-mentors/2023/01/msg00182.html).

**Comment by nbehrnd** [comment]
@rzr It is my impression that this thread can be closed because the objective has been reached.  Do you join this perspective?

A small update of the `ruby-mdl` package for Debian, `0.12.0-3` currently in `unstable` possibly enters `testing` by 2022-02-10 Fri, or 2022-02-11 Sat (cf. the [package tracker](https://tracker.debian.org/pkg/ruby-mdl)).  According to repology, some relatives of Debian (Ubuntu, Raspian, Devuan, Kali) monitor Debian closely enough that either `0.12.0-3` will equally enter their repositories, or keep `0.12.0-2`; the two versions should not differ in the functionality to run a `mdl my.md` on the CLI.  True, there are some points which can be improved; though I suggest a new thread of discussion might serve well/better, than this one.  On the other hand, I would like to refrain to submit a new package version to Debian until Q2/Q3 of 2023 / until the in charge to prepare `bookworm` to ship as `stable` were able to complete their work.

**Comment by rzr** [comment]
Yes good job: It's in testing repo:

https://tracker.debian.org/pkg/ruby-mdl

You can keep working on current version at:

https://salsa.debian.org/nbehrnd/ruby-mdl

Let's close this but feel free to comment if you want to keep us updated

**Comment by nbehrnd** [comment]
@rzr I just prepared a new .deb package (assembled on Debian 13/trixie, however equally found to be functional on a separate installation of Xubuntu 22.04 LTS Jammy) for version 0.13.0 filed on https://mentors.debian.net/package/ruby-mdl/.  As I don't know when a full Debian Developer reviews* and eventually launches the upload to Debian (the automatic checks appear reasonably fine), this could accelerate access to the functionality for you (or/and other interested, e.g. in derivatives as the Ubuntu ecosystem).

One earlier bemol of the seniors was the naming of the package alluding to libraries.  Inspired by the names [repology.org](https://repology.org/project/mdl-markdownlint/packages) lists for markdownlint in other distributions, I elected `mdl` for brevity.  This however requires to remove `ruby-mdl` prior to running `sudo dpkg -i ./mdl_0.13.0-1_all.deb`.

\* Are you (or an other reader of this thread) a Debian Developer who would consider sponsoring my advance to be a *direct maintainer* -- which would provide me permission to eventually upload updates/upgrades specific to markdownlint to Debian myself?

[mdl_0.13.0-1_all.deb.zip](https://github.com/markdownlint/markdownlint/files/12841051/mdl_0.13.0-1_all.deb.zip)

**Comment by rzr** [comment]
hi nice effort, I am not a DD but may I suggest to relocate your repo https://salsa.debian.org/nbehrnd/ruby-mdl to the ruby team and see if anyone is able to upload it . meanwhile you can apply to be become a debian maintainer then you'll be granted to upload without DD
